src/services/data-loader/extract-headings.ts

```diff
@@ -0,0 +1,34 @@
+import {
+  Heading as HeadingContent,
+  Node,
+} from '@stencila/schema';
+
+export type Content = Array<Node> | Node;
+
+export type Heading = {
+  id: string,
+  text: Content,
+};
+
+const isHeadingContent = (content: Content): content is HeadingContent => !Array.isArray(content) && typeof content === 'object' && content?.type === 'Heading';
+
+const extractHeadingContentPart = (contentPart: Content): HeadingContent[] => {
+  if (isHeadingContent(contentPart)) {
+    return [contentPart];
+  }
+
+  if (Array.isArray(contentPart)) {
+    return contentPart.flatMap(extractHeadingContentPart);
+  }
+
+  return [];
+};
+
+export const extractHeadings = (content: Content): Heading[] => {
```

Comment:
We should put some tests around this function.

Reply:
Can do, it should be noted this was ripped straight from import to make typing a little easier. Perhaps a follow up ticket is needed to test, run and store the result of this in the database when the post request is made from import.

Reply:
That's great context. The lack of tests was an issue over there too and happy for that to be a followup.

We should jump on a call with a few of us. There may be an opportunity to introduce this in a phased way so the likelihood of success in deployment is greater. If there are pieces like this that can be actioned in isolation then they are easier to approved then 3 PR's to 3 different repo's that need to be delivered together.

Reply:
It worth also pointing out that an almost identical version was the original basis for this function in import at https://github.com/elifesciences/enhanced-preprints-server/blob/87e27e254c4e1920be0437f3c501c24a4a2f0cf7/src/services/data-loader/data-loader.ts#L184-L216 . So it's kind of this functions grandparent, sat right next to it. :)

The thinking when we moved it to import was to make it so that epp-server didn't need to be aware of content schema at all - it just stores whatever it's given. That way we could remove the stencila dependencies in server and keep it "agnostic" of what is how those pieces are collected and transformed into these bits (import) and of how those are transformed into a final page (client). All it knows, is that client cares about a list of headers, separate to a (JSON) blob of content, separate to the abstract blob of content, etc. 

In fact - maybe it belongs in client after all?

Any way - we should work to resolve which is the latest/best/most-edge-cases-covered version, and make sure we use that wherever it lands.